src/main/scala/is/hail/io/bgen/BgenLoader.scala

```diff
@@ -22,18 +22,16 @@ object BgenLoader {
   def load(hc: HailContext, files: Array[String], sampleFile: Option[String] = None,
     tolerance: Double, nPartitions: Option[Int] = None): GenericDataset = {
     require(files.nonEmpty)
-    val samples = sampleFile.map(file => BgenLoader.readSampleFile(hc.hadoopConf, file))
+    val preIds = sampleFile.map(file => BgenLoader.readSampleFile(hc.hadoopConf, file))
       .getOrElse(BgenLoader.readSamples(hc.hadoopConf, files.head))
 
-    val duplicateIds = samples.duplicates().toArray
-    if (duplicateIds.nonEmpty) {
-      val n = duplicateIds.length
-      warn(
-        s"""found $n duplicate sample ${ plural(n, "ID") }
-           |  Duplicate IDs: @1""".stripMargin, duplicateIds)
+    val (sampleIds, duplicates) = mangle(preIds, "D" * _)
```

Comment:
Sorry, `f` was the name of the function in your mangle function. I prefer `_N`. There is no default. No other tools handle duplicates as far as I know -- it is extremely difficult to do. PLINK v1.7 had the behavior where it allowed duplicates for some operations. It also had strange behavior like `--remove`  would remove only the first occurrence of the ID and not any of the subsequent ones. PLINK v1.9 would just not allow duplicates at all. For Ricopili, we would have to hand-modify the .fam file before using PLINK.